### DISCUSSION TITLE
Feature comparison with django-slippers

### DISCUSSION THREAD

**Discussion by JuroOravec** [django-components/django-components#485] [Ideas]
As I was looking for inspiration for some features, I had a look at django-slippers and django-web-components, and compared the features provided by those libraries.

Here I'm documenting django-slippers features that are either 1) significantly diffrent from django-components, or 2) could be discussed/considered for django-components.

As you can see below, I marked all features as "resolved", meaning that I don't think there's much to consider here. But feel free to comment if you disagree or if I've missed something.

Feature name | Description | Resolved
---|---|---
[Defining known components in a yaml file](https://mitchel.me/slippers/docs/registering-components/) | IMO this is only a consequence of Slippers components being defined only as an HTML file, so there's nowhere to run `registry.register("my_comp", Comp)`. With django-component's you can do the same by importing the component classes in `apps.py`. Plus django-components has autodiscovery. <br/><br/> So IMO not much to consider here. | [X]
Component inputs are kwargs only | Django components accept both args and kwargs. This might become relevant once discussing [Typed component inputs](https://github.com/EmilStenstrom/django-components/issues/483). <br/><br/> Also implemented in django-web-components, so marking as resolved and instead consider this feature in comparison with `django-web-components`.  | [X]
Support both slot-ful and slot-less component tags | Recently django-components changed to support only slot-full components, so users always have to write the `{% endcomponent %}` in `{% component "comp_name" %}{% endcomponent %}`. This could be possibly revisited. <br/><br/> If so, however, I suggest to see instead the implementation of [django-web-components](https://github.com/Xzya/django-web-components/tree/master?tab=readme-ov-file#rendering-components), which is more thorough, so marking as resolved and instead consider this feature in comparison with `django-web-components`. | [X]
Shorthand component tag name | Slippers allows to use the component's registered name as component tags: <br/> `{% my_comp %}{% endmy_comp %}`. I like the idea because IMO it's clearer to read. <br/><br/> But I prefer much more the approach taken by `django-web-components`, where one can configure how to render the component tags. So users could decide if they want <br/> `{% component "my_comp" %}{% endcomponent %}` <br/> or `{% my_comp %}{% endmy_comp %}` <br/> or something else. So marking as resolved and instead consider this feature in comparison with `django-web-components`. | [X]
[HTML attribute formatting](https://mitchel.me/slippers/docs/template-tags-filters/#attrs) | Again, I prefer the approach taken by [django-web-components](https://github.com/Xzya/django-web-components#default--merged-attributes), so marking as resolved and instead consider this feature in comparison with `django-web-components`. | [X]
[`{% var %}` tag](https://mitchel.me/slippers/docs/template-tags-filters/#var) | django-components doesn't need this, because user can define any variables on the Python side in the `get_context_data` method | [X]
[`{% fragment %}` tag ](https://mitchel.me/slippers/docs/template-tags-filters/#fragment)| This tag makes sense in Slippers, because Slippers supports only default slot, whereas other slots have to be passed as variables. Basically, Slippers is closer to how React does slots, while Django-components is closer to Vue's slot handling. <br/><br/> And again this tag is needed in Slippers to some extend, because it misses a way to execute Python code per component. <br/><br/> Instead, if we need to render some component and then pass the rendered content into another component, we can do in Python, so marking this as resolved. | [X]
[`{% match %}` tag](https://mitchel.me/slippers/docs/template-tags-filters/#match) | While this is an interesting feature, in django-components users can achieve the same on Python side in `get_context_data`, so marking this as resolved. | [X]

**Comment by dylanjcastillo** [comment]
I'm not that familiar with `django-slippers` but do you know if their `fragment` allow for [template fragments](https://v1.htmx.org/essays/template-fragments/) frequently discussed in HTMX circles?

So you'd be able to do something like (copying the example verbatim from the essay):

```html
<html>
    <body>
        <div hx-target="this">
          #fragment archive-ui
            #if contact.archived
            <button hx-patch="/contacts/${contact.id}/unarchive">Unarchive</button>
            #else
            <button hx-delete="/contacts/${contact.id}">Archive</button>
            #end
          #end
        </div>
        <h3>Contact</h3>
        <p>${contact.email}</p>
    </body>
</html>
```

And then you can either render the full "component":
```python
  Contact c = getContact();
  ChillTemplates.render("/contacts/detail.html", "contact", c);
```

Or only that "fragment":
```python
  Contact c = getContact();
  ChillTemplates.render("/contacts/detail.html#archive-ui", "contact", c);
```

In `django-components`, we wouldn't be able to do this now IIUC. If that's the case, that wouldn't be on feature parity right?

**Reply by JuroOravec** [reply]
Ah, thanks for pasting the example, I think it finally clicked with me as for what the template fragments are!

But I don't think so. The slipper's `fragment` works such that when you define a fragment, the containing template is rendered, and the result is assigned to a variable. And then you can pass that rendered content as a variable.

```django
{% fragment as my_frag %}
<span>
	abc
</span>
{% endfragment %}

<div>
	{{ my_frag }}
</div>
```

Would render
```html
<div>
	<span>
		abc
	</span>
</div>
```

The equivalent with django-components would be:
```py
class MyComp(component.Component):
	template= """
		<div>
			{{ my_var }}
		</div>
	"""

	def get_context_data(self):
		my_frag = Template("""
			<span>
				abc
			</span>
		""").render()
		return {"my_frag": my_frag}

```

**Comment by EmilStenstrom** [comment]
Are we sure that we need feature parity? I'm a bit worried we will add features that no one using the library really wants...

**Reply by JuroOravec** [reply]
Not necessarily, each of the points is up for discussion. There's a few reasons why I wrote this down:
1. Mainly, I did the field work, so might as well share it.
2. This can be useful for passer-by's to assess whether they need/want this or the other library
3. Slippers does not have insignificant amount of stars (at least compared to django-components), so it may be that it is fulfilling some user needs that django-components is possibly not addressing.

**Reply by dylanjcastillo** [reply]
Yes, agreed. I don’t think feature parity should be a goal, but it always helpful to see what others in a similar space are doing. We could eventually use it as input for one of these django-components vs. X comparison tables that people like to see when picking frameworks/libraries.

**Reply by tissieres** [reply]
My 2 cents: I first settled on slippers for my components since it was the only available library available with django-components. At that time, I found django-components too verbose and I did not like being forced to split my component in separate files. I then discovered django-web-components and switch immediatly to it since it solved all my issues with slippers. In my opinion, even if I liked slippers at that time, there is not much to bring to django-components currently.

**Reply by JuroOravec** [reply]
Thanks for feedback, I've renamed the thread to "feature comparison" so it does not imply that we necessarily want to implement all of that.

Also, for the record, I made a similar comparison with [django-web-components](https://github.com/EmilStenstrom/django-components/discussions/486).

---

@tissieres Agree, it was similar with me, that at first I didn't like splitting the component to multiple files and verbosity of django-components (but I wasn't aware of django-web-components, so I stayed 😄). Maybe the django-components' [single file component](https://github.com/EmilStenstrom/django-components#using-single-file-components) feature should more prominent so it's clear that people don't necessarily need to split the component to files?

And agree also with the assessment of Slippers, it's relatively simpler featureset, and for the features that do differ, I think that considering them in the context of `django-web-components` makes more sense. `django-web-components` is practically a superset of `django-slippers` at this point, and imo does some of the features better.
